src/Language/PureScript/Errors.hs

```diff
@@ -250,12 +251,70 @@ errorMessage err = MultipleErrors [err]
 onErrorMessages :: (ErrorMessage -> ErrorMessage) -> MultipleErrors -> MultipleErrors
 onErrorMessages f = MultipleErrors . map f . runMultipleErrors
 
+
+replaceUnknowns' :: Type -> State [Unknown] Type
+replaceUnknowns' (TUnknown u) = do
+  table <- get
+  (newTable, newPos) <- return (lookupTable u table)
+  put newTable
+  return (TUnknown newPos)
+    where
+      lookupTable :: Unknown -> [Unknown] -> ([Unknown], Int)
+      lookupTable x table = case (findIndexInList (== x) table) of
+                                Nothing -> (table ++ [x], length table)
+                                Just i  -> (table, i)
+
+      findIndexInList :: (a -> Bool) -> [a] -> Maybe Int
+      findIndexInList = findIndexInListHelp 0
+
+      findIndexInListHelp _ _ [] = Nothing
+      findIndexInListHelp count f (h:t) = if (f h)
+                                          then Just count
+                                          else findIndexInListHelp (count+1) f t
+replaceUnknowns' other = return other
+
+
+replaceUnknowns :: Type -> State [Unknown] Type
+replaceUnknowns = everywhereOnTypesTopDownM replaceUnknowns'
+
+
+simplifyUnknowns :: ErrorMessage -> State [Unknown] ErrorMessage
```

Comment:
This function looks useful, but can be far more generic. See if you can parameterize over `f`, then remove the type signature and see what `ghc` infers.